File: wander/src/wander.cpp
#include "ros/ros.h"
#include "sensor_msgs/LaserScan.h"
#include "geometry_msgs/Twist.h"
#include <numeric>

using namespace std;

geometry_msgs::Twist *
wander(const sensor_msgs::LaserScan::ConstPtr& msg)
{
    assert(msg->ranges.size() >= 30);
    int angle;
    int halt = 0;
    int mid = msg->ranges.size() / 2;
    geometry_msgs::Twist * cmd = new geometry_msgs::Twist();
    // halt if an object is less than 2m in a 30deg angle
    for (angle = mid - 15; angle < mid + 15; angle++) {
        if (msg->ranges[angle] < 2) {
            halt = 1;
            break;
        }
    }
    if (halt != 0) {
        double midL, midR;
        midL = std::accumulate(msg->ranges.begin(), msg->ranges.end()-mid, 0);
        midR = std::accumulate(msg->ranges.begin()+mid, msg->ranges.end(), 0);
        // we go to the highest-range side scanned
        if (midL < midR) {
            cmd->angular.z = -1.0;
        } else {
            cmd->angular.z = +1.0;
        }
    } else {
        cmd->linear.x = 1.0;
    }

    return cmd;
}

ros::Publisher topic;

void handle_lidar(const sensor_msgs::LaserScan::ConstPtr& msg)
{
    geometry_msgs::Twist * cmd = wander(msg);
    topic.publish(*cmd);
    delete(cmd);
}

int main(int argc, char **argv)
{
    // ./bin/wander cmd:=/robot/motion laser:=/robot/sick
    ros::init(argc, argv, "wander");
    ROS_INFO("wander roscpp initilized");
    ros::NodeHandle n;

    topic = n.advertise<geometry_msgs::Twist>("cmd", 1000);
    ros::Subscriber sub = n.subscribe("laser", 1000, handle_lidar);
    ros::spin();
    return 0;
}

File: wander/nodes/wander.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import LaserScan
from geometry_msgs.msg import Twist

topic = None

def wander(message):
    assert(len(message.ranges) >= 30)
    mid = len(message.ranges) // 2
    cmd = Twist()
    # halt if an object is less than 2m in a 30deg angle
    halt = False
    for distance_to_object in message.ranges[mid-15:mid+15]:
        if distance_to_object < 2:
            halt = True
            break
    if halt:
        # we go to the highest-range side scanned
        if sum(message.ranges[:mid]) > sum(message.ranges[mid:]):
            cmd.angular.z = -1
        else:
            cmd.angular.z = +1
    else:
        cmd.linear.x = 1
    # publish twist
    topic.publish(cmd)

if __name__ == '__main__':
    # ./nodes/wander.py cmd:=/robot/motion laser:=/robot/sick
    rospy.init_node('wander')
    rospy.loginfo("wander rospy initialized")
    topic = rospy.Publisher('cmd', Twist)
    # see: ros.org/doc/api/sensor_msgs/html/msg/LaserScan.html
    rospy.Subscriber('laser', LaserScan, wander)
    rospy.spin() # block while ROS-node is up, Ctrl+C to stop
